File: backend/app/schemas/zzp_import.py
"""
ZZP Import Schemas

Pydantic schemas for CSV import preview and confirm flows
for customers, invoices, and expenses.
"""
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel


# ============================================================================
# Shared Import Schemas
# ============================================================================

class ImportRowError(BaseModel):
    """Error for a specific row during import."""
    row: int
    field: str
    message: str


class ImportPreviewRow(BaseModel):
    """A single parsed row for preview display."""
    row_number: int
    data: dict
    errors: List[ImportRowError] = []
    valid: bool = True


# ============================================================================
# Customer Import Schemas
# ============================================================================

class CustomerImportPreviewResponse(BaseModel):
    """Preview response for customer CSV import."""
    preview_rows: List[ImportPreviewRow]
    total_rows: int
    valid_rows: int
    error_rows: int
    errors: List[ImportRowError] = []
    # Store all parsed data for confirm step
    all_rows: List[dict] = []


class CustomerImportConfirmRequest(BaseModel):
    """Confirm request for customer CSV import – re-upload CSV."""
    pass


class CustomerImportConfirmResponse(BaseModel):
    """Result of confirmed customer import."""
    imported_count: int
    skipped_count: int
    total_count: int
    errors: List[ImportRowError] = []
    message: str


# ============================================================================
# Invoice Import Schemas
# ============================================================================

class InvoiceImportPreviewResponse(BaseModel):
    """Preview response for invoice CSV import."""
    preview_rows: List[ImportPreviewRow]
    total_rows: int
    valid_rows: int
    error_rows: int
    errors: List[ImportRowError] = []
    all_rows: List[dict] = []


class InvoiceImportConfirmResponse(BaseModel):
    """Result of confirmed invoice import."""
    imported_count: int
    skipped_count: int
    total_count: int
    errors: List[ImportRowError] = []
    message: str


# ============================================================================
# Expense Import Schemas
# ============================================================================

class ExpenseImportPreviewResponse(BaseModel):
    """Preview response for expense CSV import."""
    preview_rows: List[ImportPreviewRow]
    total_rows: int
    valid_rows: int
    error_rows: int
    errors: List[ImportRowError] = []
    all_rows: List[dict] = []


class ExpenseImportConfirmResponse(BaseModel):
    """Result of confirmed expense import."""
    imported_count: int
    skipped_count: int
    total_count: int
    errors: List[ImportRowError] = []
    message: str

File: src/lib/api/zzp-import.ts
// ============================================================================
// ZZP CSV Import API
// ============================================================================
// Extracted from src/lib/api.ts as part of the api.ts decomposition.
// Behavior, endpoints, and response shapes are unchanged.

import { api } from '../api'

export interface ImportRowError {
  row: number
  field: string
  message: string
}

export interface ImportPreviewRow {
  row_number: number
  data: Record<string, unknown>
  errors: ImportRowError[]
  valid: boolean
}

export interface CustomerImportPreviewResponse {
  preview_rows: ImportPreviewRow[]
  total_rows: number
  valid_rows: number
  error_rows: number
  errors: ImportRowError[]
  all_rows: Record<string, unknown>[]
}

export interface CustomerImportConfirmResponse {
  imported_count: number
  skipped_count: number
  total_count: number
  errors: ImportRowError[]
  message: string
}

export interface InvoiceImportPreviewResponse {
  preview_rows: ImportPreviewRow[]
  total_rows: number
  valid_rows: number
  error_rows: number
  errors: ImportRowError[]
  all_rows: Record<string, unknown>[]
}

export interface InvoiceImportConfirmResponse {
  imported_count: number
  skipped_count: number
  total_count: number
  errors: ImportRowError[]
  message: string
}

export interface ExpenseImportPreviewResponse {
  preview_rows: ImportPreviewRow[]
  total_rows: number
  valid_rows: number
  error_rows: number
  errors: ImportRowError[]
  all_rows: Record<string, unknown>[]
}

export interface ExpenseImportConfirmResponse {
  imported_count: number
  skipped_count: number
  total_count: number
  errors: ImportRowError[]
  message: string
}

export const zzpImportApi = {
  customers: {
    preview: async (file: File): Promise<CustomerImportPreviewResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<CustomerImportPreviewResponse>('/zzp/import/customers', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
    confirm: async (file: File): Promise<CustomerImportConfirmResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<CustomerImportConfirmResponse>('/zzp/import/customers/confirm', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
  },
  invoices: {
    preview: async (file: File): Promise<InvoiceImportPreviewResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<InvoiceImportPreviewResponse>('/zzp/import/invoices', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
    confirm: async (file: File): Promise<InvoiceImportConfirmResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<InvoiceImportConfirmResponse>('/zzp/import/invoices/confirm', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
  },
  expenses: {
    preview: async (file: File): Promise<ExpenseImportPreviewResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<ExpenseImportPreviewResponse>('/zzp/import/expenses', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
    confirm: async (file: File): Promise<ExpenseImportConfirmResponse> => {
      const formData = new FormData()
      formData.append('file', file)
      const response = await api.post<ExpenseImportConfirmResponse>('/zzp/import/expenses/confirm', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      return response.data
    },
  },
}
